File: src/methods/BFS.js
export default class BFS {
  constructor(initial, goals){
    this._initial = initial;
    this._goals = goals;
    this._path = [];
		this._visits = 0;
  }

	get visits() {
		return this._visits;
	}

	get path() {
		return this._path;
	}


  search(){
    let current = this._initial;
		let amount = [];
    while(current !== null && JSON.stringify(current.value) !== JSON.stringify(this._goals.value)){
      
      let promise = new Promise((resolve, reject) => {
        current.successors()
        .forEach(state => {
          amount.push(state);
        })

        current = amount.shift();

        resolve();
      })

      promise.then(res => this._visits++);
    }

		if(current !== null){
			this._path.push(current);

			while(current && current.parent !== null){
				current = current.parent;
				this._path.push(current);
			}
		}
    return current;
  }
}

File: Java/src/methods/BFS.java
package methods;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

import models.State;

public class BFS {
	private static State initial;
	private static State goals;
	private static ArrayList<ArrayList<Integer>> path;
	private int visits;
	
	public BFS(State initial, State goals) {
		super();
		BFS.initial = initial;
		BFS.goals = goals;
		BFS.path = new ArrayList<ArrayList<Integer>>();
		this.visits = 0;
	}


	public State getInitial() {
		return initial;
	}


	public void setInitial(State initial) {
		BFS.initial = initial;
	}


	public State getGoals() {
		return goals;
	}


	public void setGoals(State goals) {
		BFS.goals = goals;
	}


	public ArrayList<ArrayList<Integer>> getPath() {
		return path;
	}


	public int getVisits() {
		return visits;
	}

	
	public ArrayList<ArrayList<Integer>> search(){
		State current = new State(initial.getLevel(), initial.getValue(), initial.getParent(), initial.getCost());
		
		Queue<State> amount = new LinkedList<State>();
		
		 while(current != null && !current.equals(goals)){
			 ArrayList<State> successors =  current.successors();
			 
			 for(State successor : successors) {
				 amount.add(successor);
			 }
			 
			 current = amount.poll();
			 System.out.println(this.visits);
			 this.visits++;
		 }
		
		 
		 while(current != null) {
			 
			 BFS.path.add(current.getValue());
			 System.out.println(current.toString());
			 current = current.getParent();
		 }
		 
		return path;
		
	}	
	
}
